### PR TITLE
Completion mechanics haskell-completions-grab-prefix implementation

### PR DESCRIPTION
**Pull request by geraldus**
This is long story. I want to make some improvements on completion stuff.
I will add changes step by step because a bit short in time and made this PR to have some feedback and help.

### CODE REVIEW

**Comment by geraldus** [comment]
Function handles all stuff here, and also returns prefixes for pragma names itself, e.g.
`{-# OP|` ⟹ `OP`
`{-# LANG|` ⟹ `LANG`
I've merged this functionality, because both cases occurs inside `{-# #-}` comment.

Any thought about how to make it a bit readable?

**Comment by geraldus** [comment]
@gracjan updated

**Comment by geraldus** [comment]
the rest should be easier, but I want to fix presentation buffer first

File: haskell-completions.el [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+(require 'haskell-mode)
+
+(defun haskell-completions-can-grab-prefix ()
+  "Check if the case is appropriate for grabbing completion prefix.
+Returns t if point is either at whitespace character or at line
+end and preceeding character is not a whitespace or new line,
+otherwise returns nil."
+  ;; TODO Discuss this. In some cases it is better to provide completions if
+  ;; point is not at whitespace, for example when point is at closing
+  ;; parenthesis: (ma|) - point is after `a` character.
```
Comment (geraldus) [inline haskell-completions.el]: In some cases it is better to provide completions if point is not at whitespace, for example when point is at closing parenthesis: `(ma|)` - point is after `a` character. Any thoughts?

**Comment by geraldus** [comment]
Well, I believe I've covered all prefix cases originally defined here<sup>[1]</sup>.  Now I can write a general function for grabbing prefix at point which first invokes pragma-prefix function and if its result is nil it returns the result of identifier-prefix function.

And after that I can update completions at point function: split it to synchronous and asynchronous ones.
Synchronous function will operate like existing CAP-function, and asynchronous can provide all completions using REPL, i.e. provide completions for GHC options and language extensions running `:complete repl ":set <prefix>"`.

Any thoughts so far?

[1]:https://github.com/haskell/haskell-mode/blob/master/haskell.el#L66-L100

**Comment by gracjan** [comment]
Sound like a very good idea!

**Comment by geraldus** [comment]
A little progress here

**Comment by gracjan** [comment]
Are there some commits here that you could spin-off as a separate pull request? This is getting rather big...

**Comment by geraldus** [comment]
@gracjan I think I'm done with prefixes, can you review everything that done so far and it is could be merged after squashing some commits.  The next part is to update completion functions (still some research needed).

I have no strong opinion on some key points though:

1. Currently, all grabbing functions return `nil` if mark is activated.
2. Should completion list be provided when typing in the middle of word?

File: haskell-completions.el [hunk 2]
```diff
@@ -0,0 +1,189 @@
+;;; haskell-completions.el --- Haskell Completion package
+
+;; FIXME copyright (c) 2015 Haskell mode community. All rights reserved.
```
Comment (geraldus) [inline haskell-completions.el]: Is it okay to reserve copy rights to community rather than a person?

File: haskell-completions.el [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+;; Authors:
```
Comment (geraldus) [inline haskell-completions.el]: Originally, I've copied this from existing package, and replaced sigular `Author` to plural `Authors`. Not sure, is it right?

File: haskell-completions.el [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+;; You should have received a copy of the GNU General Public License
+;; along with GNU Emacs; see the file COPYING.  If not, write to
+;; the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+;; Boston, MA 02110-1301, USA.
+
+;;; Commentary:
+
+;; This package provides functionality for all possible auto completions for
+;; Haskell Mode.  FIXME better description.
```
Comment (geraldus) [inline haskell-completions.el]: Just a reminder not to forget to update package description

File: tests/haskell-completions-tests.el [hunk 5]
```diff
@@ -0,0 +1,366 @@
+;;; haskell-completions-tests.el --- Tests for Haskell Completion package
+
+;; FIXME copyright (c) 2015 Haskell mode community. All rights reserved.
+
+;; Authors:
+;; Arthur Fayzrakhmanov <heraldhoi@gmail.com>
```
Comment (geraldus) [inline tests/haskell-completions-tests.el]: Same copyright and authors comments as above.

Comment (gracjan) [inline tests/haskell-completions-tests.el]: For licensing discussion see #630.

**Comment by geraldus** [comment]
@gracjan I think GPL licence itself (`COPYING` file) could be spinned-off.  Also if GPL is the preferrable licensing method I would be nice to have some notice about licensing on [Contributing][contrib-wiki] page.

[contrib-wiki]:https://github.com/haskell/haskell-mode/wiki/Contributing

**Comment by gracjan** [comment]
Overall this looks good to me. Tests provide assurance that this also works.

License stuff got merged in another commit, so it can go away. Please squash what is there to squash.

(Note: when you commit you can use `git commit --squash COMMITID` that will generate specially formatted commit message for commits that are supposed to be squashed.)

**Comment by geraldus** [comment]
@gracjan I will tidy up things this evening, however, I'm still have no strong opinion about completions in the middle of strings.  Initially I've decided not to grab prefixes in the middle of string (hence no completions in this case) being influenced by some comments in **company-mode** sources<sup>[1][company-comments]</sup>.  But, when I write Elisp code **company-mode** works even in the mid of the word and it feels quite comfortable and useful, so this was wrong decision from the very beginning and provide no completions in the middle of string is likely a misfeature.

Now I'm leaning to other direction, but I don't know and can't recall how completions provision works in other packages (like **auto-complete**) and/or in other editors/IDEs  _(I can check how Sublime works in the evening).

[company-comments]:https://github.com/company-mode/company-mode/blob/master/company.el#L340-L347

**Comment by gracjan** [comment]
I would love completions to work inside comments!

Note that QuasiQuotes are done using strings, so having completions there
can be also useful.

I vote for having completions everywhere!

**Comment by geraldus** [comment]
OK, I've tested Sublime and it does not completes in the middle of string, so does **company-mode** with **elisp** code.  But them both provide completions when point is at punctuation, e.g. `co|)`, `com|,`, `comp|.`, and etc, and we really need it too for **haskell-mode** _(currently it's not supported)_.

I also want to have completions inside comments and string literals, this is one of the reasons of my efforts on completion improvements.  Currently, it is possible to grab prefix inside comments and strings, but I think I need to implement special completion type for these cases, e.g. `'haskell-completions-literal-prefix`.  I plan to make some hacking around `dabbrev-expand` function to get completions in this case, and as a free bonus we'll have a completions for local identifiers (in `let` blocks) and newly added things with unsaved buffer. 
If I'm not mistaken REPL does not provide completions for local identifiers, also it is required to reload file to have completions for things added since last load.  Similarly, it is required to save buffer when completions come from **TAGS** file.

As for quasi-quotes, my current haskell workflow have lot of them, so if it's possible to make QQs support a bit better I'd love to have it.  But I have nothing to say about this stuff yet.

**Comment by geraldus** [comment]
pinging @gracjan, please review carefully final update

**Comment by gracjan** [comment]
Code is nice, can be merged as is. Do you anything more you want to do about this?

How can I see this working besides of unit tests?

**Comment by geraldus** [comment]
@gracjan nope, this is supposed to be merged as is, the next PR will touch completion functionality itself.  I'm testing this stuff manually running `M-: (haskell-completions-grab-prefix)`, etc. at various places of arbitrary haskell file.

**Pull request merged**